README.md
# Test & Publish to Cloudflare


![Tests](https://github.com/carlosdevpereira/test-publish-to-cloudflare/actions/workflows/run-unit-tests.yml/badge.svg)
![Coverage](https://img.shields.io/badge/Coverage-90&#65130;-brightgreen)
![Tests](https://img.shields.io/badge/Tests-%E2%9C%94%2042%20%7C%20%E2%9C%98%200%20%7C%20%E2%8C%80%200-brightgreen)

> This is a Github action that runs the unit tests of a repository, generates a test coverage report, uploads the report to Cloudflare Pages and comments the results on available pull requests.

## Requirements:

1. A Javascript repository, hosted on Github, using Jest for unit testing
2. A Cloudflare account with access to Cloudflare Pages and:
   - A Cloudflare API Token
   - The Cloudflare Account ID
   - A Cloudflare Pages project (to upload coverage reports)

## How to use:

In your repository, add the following step to your branch pushes workflow:

```yaml
    - name: Test & Publish to Cloudflare 🧪
        uses: carlosdevpereira/test-publish-to-cloudflare@v1
        with:
          githubToken: ${{ secrets.GITHUB_TOKEN }}
          cloudflareProjectName: THE_NAME_OF_YOUR_CLOUDFLARE_PROJECT
          cloudflareApiToken: ${{ secrets.CLOUDFLARE_API_TOKEN }}
          cloudflareAccountId: ${{ secrets.CLOUDFLARE_ACCOUNT_ID }}
          baseCloudflareDeploymentUrl: THE_NAME_OF_YOUR_CLOUDFLARE_PROJECT.pages.dev
```

### Example workflow

```yaml
name: Build

on:
  push:
    branches:
      - "**"

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: "16"
          cache: "npm"

      - name: Install dependencies 📦
        run: npm ci

      - name: Test & Publish to Cloudflare 🧪
        uses: carlosdevpereira/test-publish-to-cloudflare@v1
        with:
          githubToken: ${{ secrets.GITHUB_TOKEN }}
          cloudflareProjectName: THE_NAME_OF_YOUR_CLOUDFLARE_PROJECT
          cloudflareApiToken: ${{ secrets.CLOUDFLARE_API_TOKEN }}
          cloudflareAccountId: ${{ secrets.CLOUDFLARE_ACCOUNT_ID }}
          baseCloudflareDeploymentUrl: THE_NAME_OF_YOUR_CLOUDFLARE_PROJECT.pages.dev

      - name: Build app
        run: npm run build
```

### Expected parameters:

- **framework**: Defines which testing framework this action should use to run the tests of your repository. For now the only valid value is `jest` but `vitest` support will be added soon.

- **githubToken**: The token that this action should use to authenticate requests to the Github API.

- **cloudflareApiToken**: The API Token generated on Cloudflare. Ideally this token should be saved in the repository secrets and shoudn't be shared.

- **cloudflareAccountId**: The Account ID of the Cloudflare account that matches the API Token passed in. This parameter should also be saved in a repository secret.

- **cloudflareProjectName**: The name of the project on Cloudflare Pages that will receive the coverage reports.

- **baseCloudflareDeploymentUrl**: The base URL for deployment previews of your Cloudflare Pages project. Example: If you create a project named `my-awesome-project` on Cloudflare Pages, by default, the base URL will be: `my-awesome-project.pages.dev`. This URL can also be a custom domain defined in Cloudflare.


CHANGELOG.md
# Changelog

## [1.9.0](https://github.com/carlosdevpereira/test-publish-to-cloudflare/compare/v1.8.0...v1.9.0) (2022-12-18)


### 👷 Maintenance

* add badges to README.md ([1e948f5](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/1e948f5d9772325e33c3b3915079e8e2dcc7f52c))
* add example workflow to README.md ([9c4bd05](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/9c4bd0541d1239a42712842f0abe68fc8e132ead))
* add global jest setup ([1068c03](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/1068c030de0f2ecbe3f3022866bee34ab889cf22))
* add infos on upload ([ba46bd5](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/ba46bd5b71401bdbd65a3dd881ab13db7328e2d2))
* add jsconfig with folder alias and refactor tests ([994604c](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/994604cffbc060b4210c4b6fa1e12b7bb0e0e793))
* add node_modules ([26ce6d4](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/26ce6d40b235104e891c0b3515597f6ba787474f))
* declare as required only the required inputs ([d95a7ba](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/d95a7ba459e85322fba4f0c05816eaa07bbc1404))
* get branch that triggered workflow from context ([c90ef4a](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/c90ef4a8a177ec4d4af880af8b22ffa071b19c86))
* get head results from uploaded report ([f5e803c](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/f5e803cb42fe41305cdeae38f36527c636fe25c0))
* get head results from uploaded report ([34eda53](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/34eda53fe39e55058aa7847b54be5eaf5cc37745))
* get head results from uploaded report ([46a61e6](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/46a61e60cf6c16f51374cf5967a9fb30d48eaecb))
* get head results from uploaded report ([29550cb](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/29550cb7655eed82fede4c196cd148192c1d2659))
* get head results from uploaded report ([c768e51](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/c768e51259c869f4e0d2dc5b47b9b42cb9193512))
* get head results from uploaded report ([4679ef9](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/4679ef9077a2f18072bb922dc0ef4268ef791fae))
* group octokit requests ([c9183b9](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/c9183b944eb7da861d981a5de6cc2dc0dda19a22))
* ignore dist eslint format ([a63a5f6](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/a63a5f62721e7de96c61140512c7f5f60782093b))
* ignore stderr ([c71a2e3](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/c71a2e3e4302f882c2ac057896b1e4626cf2018c))
* improve formatter rules ([d23352e](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/d23352e1388e2a1e49330b7a1ceee6e369f30407))
* increment action version ([b1f4331](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/b1f4331ae6b5ca171483a2a5b368fda759fc5123))
* log context ([fb32012](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/fb32012431e3e32cc16dad9b32e8fe34f7647233))
* log readfile ([0e95e17](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/0e95e17aa338863b33915560f548d383da0bcbe9))
* log readfile ([c9de541](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/c9de54166dc0741bf48748e86d2d0d95d09217ac))
* log search query ([c23a5ce](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/c23a5ce94453cf1181db1e121aa649527049174f))
* migrate to pnpm ([999b20f](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/999b20f8b96da92bbd9c5be8490d60a4a1cb15b4))
* print action progress to github console ([6168dc7](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/6168dc78d583154c48bec1b2cd228f30e56373e2))
* print to github console available prs and outputs ([cd200b0](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/cd200b0dabec34f663323e11cea2e25a22bcd907))
* re-add node_modules as stated by Github docs ([2e32db4](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/2e32db43789231fd50f61c6e95289ba47c0b962c))
* refactor action logic ([ccb0323](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/ccb032390014c8c60572693d3b5d383e4a724b58))
* refactor code to reuse results from head commit ([7d4b39f](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/7d4b39f570fe1f3f27dc5d06718f64b8512850f1))
* refactor config ([b0d0d7a](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/b0d0d7a525fd8048f831ee0fe8c004b9639696fa))
* remove branchName from readme ([e38dcf0](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/e38dcf05cb6fd8d6f6e877fff62a46fe67b356e2))
* remove dist folder again ([6910692](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/6910692b50d62f3fd5e9374bb1014a9412113182))
* remove node_modules and dist / add release process ([1bc874d](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/1bc874d00890c2c8d8d2b6920822eb687f7a7edb))
* rename workflow ([cea31f0](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/cea31f0d1b181891718f09ce4555d4e515645d50))
* rollback to commonjs syntax for GH compatibility ([48c8363](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/48c8363111e7b7f1c0120d5925595b298a80d111))
* set new release version ([5d425bf](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/5d425bffd94ced733b07e1afdc0e35767edb4a90))
* use fs to write file instead of exec ([5861b22](https://github.com/carlosdevpereira/test-publish-to-cloudflare/commit/5861b22fcd1475b9bebfce998398daff0047f45f))

src/Framework.js
const fs = require('node:fs');
const { exec } = require('@actions/exec');

const SUPPORTED_TEST_FRAMEWORKS = ['jest'];
const COVERAGE_OUTPUT_FOLDER = './coverage';

class Framework {
  constructor(frameworkName) {
    if (!SUPPORTED_TEST_FRAMEWORKS.includes(frameworkName)) {
      throw new Error(
        'Unsupported test framework selected. Valid options are: '
        + SUPPORTED_TEST_FRAMEWORKS.join(', ')
      );
    }

    this.name = frameworkName;
    this.testResults = null;
  }

  /**
   * Runs the unit tests for the selected test framework.
   *
   * (If you use a different framework, and would like
   * to use this action, feel free to open a feature request
   * in this repository 😉).
   */
  async runTests(experimentalModules = false) {
    const JEST_PATH = `node ${experimentalModules ? '--experimental-vm-modules' : ''} ./node_modules/jest/bin/jest.js`;
    const JEST_FLAGS = '--no-cache --detectOpenHandles --coverage --json';

    let testRunStats = '';
    await exec(`${JEST_PATH} ${JEST_FLAGS}`, undefined, {
      listeners: {
        stdout: (data) => {
          testRunStats += data.toString();
        },
      },
    });

    const coverageSummary = fs.readFileSync(`${COVERAGE_OUTPUT_FOLDER}/coverage-summary.json`, 'utf-8');

    this.testResults = parseTestResults(
      JSON.parse(testRunStats),
      JSON.parse(coverageSummary)
    );

    return this.testResults;
  }
}

const parseTestResults = (testRunStats, coverageSummary) => {
  const percentage = Math.floor([
    coverageSummary.total.lines.pct,
    coverageSummary.total.statements.pct,
    coverageSummary.total.functions.pct,
    coverageSummary.total.branches.pct,
  ].reduce((a, b) => {
    return a + b;
  }, 0) / 4);

  const summary = {
    statements: {
      total: coverageSummary.total.statements.total,
      covered: coverageSummary.total.statements.covered,
      skipped: coverageSummary.total.statements.skipped,
      percentage: coverageSummary.total.statements.pct,
    },
    lines: {
      total: coverageSummary.total.lines.total,
      covered: coverageSummary.total.lines.covered,
      skipped: coverageSummary.total.lines.skipped,
      percentage: coverageSummary.total.lines.pct,
    },
    functions: {
      total: coverageSummary.total.functions.total,
      covered: coverageSummary.total.functions.covered,
      skipped: coverageSummary.total.functions.skipped,
      percentage: coverageSummary.total.functions.pct,
    },
    branches: {
      total: coverageSummary.total.branches.total,
      covered: coverageSummary.total.branches.covered,
      skipped: coverageSummary.total.branches.skipped,
      percentage: coverageSummary.total.branches.pct,
    },
  };

  const stats = {
    startTime: testRunStats.startTime,
    endTime: testRunStats.testResults[testRunStats.testResults.length - 1].endTime,
    testSuites: {
      total: testRunStats.numTotalTestSuites,
      passed: testRunStats.numPassedTestSuites,
      failed: testRunStats.numFailedTestSuites
    },
    tests: {
      total: testRunStats.numTotalTests,
      passed: testRunStats.numPassedTests,
      failed: testRunStats.numFailedTests
    },
    snapshots: {
      total: testRunStats.snapshot.total
    }
  };

  return {
    coverage: {
      summary,
      percentage,
      badges: {
        coverage: createCoverageBadge(percentage),
        tests: createTestsBadge(stats.tests),
      }
    },

    stats
  };
};

const createCoverageBadge = (percentage) => {
  let color = 'lightgrey';

  if (percentage >= 90) color = 'brightgreen';
  else if (percentage >= 75) color = 'green';
  else if (percentage >= 50) color = 'yellowgreen';
  else if (percentage >= 35) color = 'yellow';
  else color = 'red';

  return `https://img.shields.io/badge/Coverage-${percentage}&#65130;-${color}`;
};

const createTestsBadge = (tests) => {
  let color = 'lightgrey';

  if (tests.total === tests.passed) color = 'brightgreen';
  else if (tests.failed > 0) color = 'red';
  else color = 'yellow';

  return `https://img.shields.io/badge/Tests-%E2%9C%94%20${tests.passed}%20%7C%20%E2%9C%98%20${tests.failed}%20%7C%20%E2%8C%80%20${tests.total - tests.passed - tests.failed}-${color}`;
};

module.exports = Framework;